paddle/fluid/imperative/reducer.h

```diff
@@ -0,0 +1,224 @@
+// Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <algorithm>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+#include "paddle/fluid/framework/data_type.h"
+#include "paddle/fluid/imperative/layer.h"
+#include "paddle/fluid/imperative/variable_wrapper.h"
+#include "paddle/fluid/memory/memory.h"
+
+#if defined(PADDLE_WITH_NCCL)
+#include "paddle/fluid/imperative/all_reduce.h"
+#include "paddle/fluid/operators/math/concat_and_split.h"
+#include "paddle/fluid/operators/strided_memcpy.h"
+#include "paddle/fluid/platform/cuda_resource_pool.h"
+#endif
+
+namespace paddle {
+namespace imperative {
+
+#if defined(PADDLE_WITH_NCCL)
+template <typename T>
+void ConcatTensorsForAllReduce(
+    const platform::CUDADeviceContext& context,
+    const std::vector<framework::Tensor>& dense_tensors,
+    framework::Variable* p_dense_contents) {
+  operators::math::ConcatFunctor<platform::CUDADeviceContext, T>
+      concat_functor_;
+  concat_functor_(context, dense_tensors, 0,
+                  p_dense_contents->GetMutable<framework::LoDTensor>());
+}
+
+template <typename T>
+void SplitTensorsForAllReduce(const platform::CUDADeviceContext& context,
+                              framework::Variable* p_dense_contents,
+                              std::vector<framework::Tensor>* p_dense_tensors) {
+  auto* in = p_dense_contents->GetMutable<framework::LoDTensor>();
+  std::vector<framework::Tensor*> outs;
+  std::vector<const framework::Tensor*> shape_refer;
+
+  outs.reserve(p_dense_tensors->size());
+  shape_refer.reserve(p_dense_tensors->size());
+
+  for (auto& tensor : *p_dense_tensors) {
+    outs.emplace_back(&tensor);
+    shape_refer.emplace_back(&tensor);
+  }
+  // Sometimes direct copies will be faster
+  if (p_dense_tensors->size() < 10) {
+    operators::StridedMemcpyWithAxis0<T>(context, *in, shape_refer, &outs);
+  } else {
+    operators::math::SplitFunctor<platform::CUDADeviceContext, T>
+        split_functor_;
+    split_functor_(context, *in, shape_refer, 0, &outs);
+  }
+}
+
+struct Group {
```

Comment:
why use struct? class is better? styleguid is
![image](https://user-images.githubusercontent.com/22561442/100351879-1dfe6080-3027-11eb-952d-c6c3c9fec9f2.png)